### PR TITLE
[16.0][IMP] purchase_vendor_promotion: automatically assign supplier with promotion

### PR DESCRIPTION
**Pull request by trisdoan**
### Note
- depends on: purchase_vendor_promotion (pending merge)

### CODE REVIEW

**Review by nilshamerlinck — CHANGES_REQUESTED** [review]
any room to improve tests to check default?

**Comment by trisdoan** [comment]
Hi anh @nilshamerlinck, I added the test :pray:

File: purchase_vendor_promotion/models/stock_route.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import fields, models
+
+
+class StockRoute(models.Model):
+    _inherit = "stock.route"
+
+    is_force_default_vendor = fields.Boolean(default=False)
```
Comment (nilshamerlinck) [inline purchase_vendor_promotion/models/stock_route.py]: I would suggest more explicit name: `force_vendor_with_best_promotion`

Would also need a `data/stock_route.xml` file to update our Buy route:

```
        <record id="purchase_stock.route_warehouse0_buy" model="stock.route">
            <field name="force_vendor_with_best_promotion">True</field>
        </record>
   ```

File: purchase_vendor_promotion/models/product_supplierinfo.py [hunk 2]
```diff
@@ -22,3 +22,13 @@ def _check_promotion_dates(self):
                     raise ValidationError(
                         _("Promotion start date must be before end date.")
                     )
+
+    def _is_promotion_active(self, date=None):
```
Comment (nilshamerlinck) [inline purchase_vendor_promotion/models/product_supplierinfo.py]: card says to also consider future promotions (even if very far)

File: purchase_vendor_promotion/models/stock_warehouse_orderpoint.py [hunk 3]
```diff
@@ -12,6 +12,20 @@ class StockWarehouseOrderpoint(models.Model):
...
+
+    @api.depends("route_id", "route_id.is_force_default_vendor")
+    def _compute_supplier_id(self):
+        for rec in self:
+            if rec.route_id and rec.route_id.is_force_default_vendor:
+                suppliers = rec.product_id._prepare_sellers(False).filtered(
+                    lambda x: x.is_promotion and x._is_promotion_active()
+                )
+                suppliers = suppliers.sorted(key=lambda x: x.price)
+                if suppliers:
```
Comment (nilshamerlinck) [inline purchase_vendor_promotion/models/stock_warehouse_orderpoint.py]: good, but card also says to select first vendor if there is none with a promotion

**Review by nilshamerlinck — CHANGES_REQUESTED** [review]
additional small requirement to make the module slightly better:
In `Purchase > Configuration > Vendor Pricelists` (product.supplierinfo), add Filter: Promotion (`is_promotion = True`)

File: purchase_vendor_promotion/models/stock_route.py (hunk 1 above)
Comment (trisdoan) [reply purchase_vendor_promotion/models/stock_route.py]: I guess you suggest to add the data file to allo's local module right (like `local-src/allo_data_stock`) ?

Or directly to this oca module ?

Comment (nilshamerlinck) [reply purchase_vendor_promotion/models/stock_route.py]: hi, I think it makes sense to add here directly as it goes well with the global purpose of the module

Comment (trisdoan) [reply purchase_vendor_promotion/models/stock_route.py]: I agree, with explicit description maybe to let users aware of this feature

**Comment by trisdoan** [comment]
Hi anh @nilshamerlinck, I updated as you suggest.

### This change
- Migration file is needed as Buy route is marked as noupdate in [here](https://github.com/odoo/odoo/blob/16.0/addons/purchase_stock/data/purchase_stock_data.xml)
- As consequence, I increased the version of  module to execute the post-migrate script. Although the module is not merged to main branch, the module is already used in allo's codebase as pending merge. Do you think we have another better option :thinking:

File: purchase_vendor_promotion/models/product_supplierinfo.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                if not rec.date_start or not rec.date_end:
+                    raise ValidationError(
+                        _("Vendor promotion requires start and end dates.")
+                    )
+                if rec.date_start > rec.date_end:
+                    raise ValidationError(
+                        _("Promotion start date must be before end date.")
+                    )
+
+    def _is_promotion_active(self, date=None):
```
Comment (nilshamerlinck) [inline purchase_vendor_promotion/models/product_supplierinfo.py]:
```suggestion
    def _is_promotion_active_or_upcoming(self, date=None):
```

File: purchase_vendor_promotion/models/product_supplierinfo.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        """
+        Consider Promotion is active if it's within start and end date or it's in the future
+        """
+        self.ensure_one()
+        if not date:
+            date = fields.Date.today()
+        if not self.date_start:
+            return self.date_end >= date
+        if not self.date_end:
+            return date <= self.date_start or date >= self.date_start
```
Comment (nilshamerlinck) [inline purchase_vendor_promotion/models/product_supplierinfo.py]:
```suggestion
```

File: purchase_vendor_promotion/models/product_supplierinfo.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+                    raise ValidationError(
+                        _("Promotion start date must be before end date.")
+                    )
+
+    def _is_promotion_active(self, date=None):
+        """
+        Consider Promotion is active if it's within start and end date or it's in the future
+        """
+        self.ensure_one()
+        if not date:
```
Comment (nilshamerlinck) [inline purchase_vendor_promotion/models/product_supplierinfo.py:33]:
```suggestion
        if not self.is_promotion:
            return False
        if not date:
```

File: purchase_vendor_promotion/models/product_supplierinfo.py (hunk 5 above)
Comment (nilshamerlinck) [reply purchase_vendor_promotion/models/product_supplierinfo.py]: we have a constraint on date_start and date_end, so we can simplify

File: purchase_vendor_promotion/readme/DESCRIPTION.rst [hunk 7]
```diff
@@ -0,0 +1,2 @@
+This module allows you to visualise the start and end date of a vendor promotion during replenishment.
+A vendor price is considered as a promotion if product supplier info is flagged as a `promotion`. Besides, reodering rules are automatically assigned with the vendor promotion. By default, Buy Route is marked to assign vendor in replenishment.
```
Comment (nilshamerlinck) [inline purchase_vendor_promotion/readme/DESCRIPTION.rst]:
```suggestion
This module allows you to visualize the start and end date of a vendor promotion during replenishment, in `Inventory > Operations > Replenishment` or `Inventory > Configuration > Products > Reordering Rules`.
A vendor price is considered as a promotion if product supplier info is flagged as a `promotion`.
If "Buy" is chosen as the Preferred Route, the vendor having the best active or upcoming promotion for the product will be automatically chosen. If no promotion is available, then first vendor will be chosen.
```

File: purchase_vendor_promotion/migrations/16.0.1.1.0/post-migrate.py
Comment (nilshamerlinck) [inline purchase_vendor_promotion/migrations/16.0.1.1.0/post-migrate.py:1]: given that it's not been merged yet on oca side, better do that as a song in the alloboissons codebase

**Review by nilshamerlinck — CHANGES_REQUESTED** [review]

File: purchase_vendor_promotion/models/product_supplierinfo.py (hunk 6 above)
Comment (trisdoan) [reply purchase_vendor_promotion/models/product_supplierinfo.py:33]: I agree with this. To respect DRY, I will remove the check used in filter method

```
promotion_suppliers = suppliers.filtered(
                    lambda x: x.is_promotion and x._is_promotion_active()
                )
```

**Comment by trisdoan** [comment]
Hi anh @nilshamerlinck, I updated as you suggest

**Review by nilshamerlinck — APPROVED** [review]

**Pull request closed without merging**